bus-mapping/src/circuit_input_builder.rs

```diff
@@ -809,6 +808,210 @@ impl<'a> CircuitInputStateRef<'a> {
+fn gen_state_access_trace<TX>(
+    _block: &eth_types::Block<TX>,
+    tx: &eth_types::Transaction,
+    geth_trace: &GethExecTrace,
+) -> Result<Vec<Access>, Error> {
+    use AccessValue::{Account, Code, Storage};
+    use RW::{READ, WRITE};
+
+    let mut call_address_stack = vec![tx.from];
+    let mut accs = vec![Access::new(None, WRITE, Account { address: tx.from })];
+    if let Some(to) = tx.to {
+        accs.push(Access::new(None, WRITE, Account { address: to }));
+        // Code may be null if the account is not a contract
+        accs.push(Access::new(None, READ, Code { address: to }));
+    } else {
+        let address = get_contract_address(tx.from, tx.nonce);
+        accs.push(Access::new(None, WRITE, Account { address }));
+        accs.push(Access::new(None, WRITE, Code { address }));
+    }
+
+    for (index, step) in geth_trace.struct_logs.iter().enumerate() {
+        let next_step = geth_trace.struct_logs.get(index + 1);
+        let i = Some(index);
+        let sender = call_address_stack[call_address_stack.len() - 1];
+        match step.op {
+            OpcodeId::SSTORE => {
+                let address = sender;
+                let key = step.stack.nth_last(0)?;
+                accs.push(Access::new(i, WRITE, Storage { address, key }));
+            }
+            OpcodeId::SLOAD => {
+                let address = sender;
+                let key = step.stack.nth_last(0)?;
+                accs.push(Access::new(i, READ, Storage { address, key }));
+            }
+            OpcodeId::SELFBALANCE => {
+                accs.push(Access::new(i, READ, Account { address: sender }));
+            }
+            OpcodeId::CODESIZE => {
+                accs.push(Access::new(i, READ, Code { address: sender }));
+            }
+            OpcodeId::CODECOPY => {
+                accs.push(Access::new(i, READ, Code { address: sender }));
+            }
+            OpcodeId::BALANCE => {
+                let address = step.stack.nth_last(0)?.to_address();
+                accs.push(Access::new(i, READ, Account { address }));
+            }
+            OpcodeId::EXTCODEHASH => {
+                let address = step.stack.nth_last(0)?.to_address();
+                accs.push(Access::new(i, READ, Account { address }));
+            }
+            OpcodeId::EXTCODESIZE => {
+                let address = step.stack.nth_last(0)?.to_address();
+                accs.push(Access::new(i, READ, Code { address }));
+            }
+            OpcodeId::EXTCODECOPY => {
+                let address = step.stack.nth_last(0)?.to_address();
+                accs.push(Access::new(i, READ, Code { address }));
+            }
+            OpcodeId::SELFDESTRUCT => {
+                accs.push(Access::new(i, WRITE, Account { address: sender }));
+                let address = step.stack.nth_last(0)?.to_address();
+                accs.push(Access::new(i, WRITE, Account { address }));
+            }
+            OpcodeId::CREATE => {
```

Comment:
When `CREATE` fails due to `ContractAddressCollision`, we might also want to add `Account` read with collided `address` into list.

Reply:
I agree that this would be desirable, but I'm not sure how to proceed here because:
1. If CREATE or CREATE2 fail, the address where the code would be deployed in the successful case can't be found in the trace (because the result is `0`)
2. For CREATE2 we can easily calculate the address from the parameters, but for CREATE we need to know the nonce; but in this function we don't yet have access to the state, (we're actually trying to get a list of accesses to know which state keys are needed).

The only thing that comes to my mind right now is to delay generating this particular Access for the step where the circuit inputs are built.  At that point we do have a copy of the state and can query the nonce to calculate the address.  So the flow could go like this
1. Query geth for Block, Txs and TxExecTraces
2. Get Accesses from TxExecTraces
3. Query geth for all accounts, storage keys, and codes from Accesses
4. Build a partial StateDB from 3
5. For each step in TxEcexTracess, gen the associated ops and all the circuit inputs; and output read accesses for accounts on failed CREATE* due to collision address
6. Query geth for extra accounts from 5
7. Merge results from 3 and 6 into a Partial State Trie to generate proofs for the circuit.

Does this make sense?  Or do you have any other idea?

Reply:
I see, for `CREATE` it's more tricky to do.

The flow makes sense to me, was just thinking step 2 aims to query all information, but it seems hard to do, then having some extra query while generating associated ops totally makes sense.